src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,188 @@
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @FunctionalInterface
+    public interface SamVoidFunction {
+        void apply() throws ApiException, InterruptedException;
+    }
+
+    @FunctionalInterface
+    public interface SamFunction<R> {
+        R apply() throws ApiException, InterruptedException;
+    }
+
+
+    static <T> T retry(SamFunction<T> function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        return samRetry.perform(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static <T> T retry(SamFunction<T> function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        return samRetry.perform(function);
+    }
+
+    static void retry(SamVoidFunction function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        samRetry.performVoid(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static void retry(SamVoidFunction function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        samRetry.performVoid(function);
+    }
+
+    private <T> T perform(SamFunction<T> function) throws InterruptedException {
+        while (true) {
+            try {
+                return function.apply();
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void performVoid(SamVoidFunction function) throws InterruptedException {
+        while (true) {
+            try {
+                function.apply();
+                return;
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void handleApiException(ApiException ex) {
+        ErrorReportException rex = convertSAMExToCommonEx(ex);
+        if (!(rex instanceof InternalServerErrorException)) {
+            throw rex;
+        }
+        logger.info("SamRetry: caught retry-able exception: " + ex);
+        // if our operation timeout will happen before we wake up from our next retry
+        // sleep, then we give up and re-throw.
+        if (operationTimeout.minusSeconds(retrySeconds).isBefore(now())) {
+            logger.error("SamRetry: operation timed out after " + operationTimeout.toString());
+            throw rex;
+        }
```

Comment:
optional style: it's a little surprising to me that this function checks the timeout and throws. consider making a private variable `lastException` and handle throwing if we're going to timeout in `sleepBeforeRetrying`. Or update the function name.

Reply:
Refactored